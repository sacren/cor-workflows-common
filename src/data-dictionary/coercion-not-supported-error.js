export default function CoercionNotSupportedError (
  baseType,
  requestedCoerceType,
  valueToCoerce
) {
  Error.captureStackTrace(this, this.constructor)

  this.message = `Cannot coerce ${baseType} into ${
    requestedCoerceType
  }: Operation not supported.`
  this.baseType = baseType
  this.requestedCoerceType = requestedCoerceType
  this.valueToCoerce = valueToCoerce

  this.name = 'CoercionNotSupportedError'
}

CoercionNotSupportedError.prototype = Object.create(Error.prototype)
CoercionNotSupportedError.prototype.constructor = CoercionNotSupportedError
