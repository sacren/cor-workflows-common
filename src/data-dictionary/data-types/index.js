import {
  BOOLEAN,
  CATEGORY,
  FORM,
  GROUP,
  NUMBER,
  ROLE,
  SUBFLOW,
  TEXT,
  USER,
  OPTIONS,
  ALL
} from '../return-types'
import * as boolean from './boolean'
import * as category from './category'
import * as form from './form'
import * as text from './text'
import * as user from './user'

export default {
  [BOOLEAN]: boolean,
  [CATEGORY]: category,
  [FORM]: form,
  [USER]: user,
  [TEXT]: text
}
