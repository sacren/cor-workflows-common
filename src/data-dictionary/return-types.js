/* Copyright Â© 2016 Kuali, Inc. - All Rights Reserved
 * You may use and modify this code under the terms of the Kuali, Inc.
 * Pre-Release License Agreement. You may not distribute it.
 *
 * You should have received a copy of the Kuali, Inc. Pre-Release License
 * Agreement with this file. If not, please write to license@kuali.co.
 */
export const BOOLEAN = 'boolean'
export const CATEGORY = 'category'
export const CATEGORY_LIST = 'category-list'
export const FORM = 'form'
export const GROUP = 'group'
export const GROUP_LIST = 'group-list'
export const NUMBER = 'number'
export const ROLE = 'role'
export const SUBFLOW = 'subflow'
export const TEXT = 'text'
export const TEXT_LIST = 'text-list'
export const USER = 'user'
export const OPTIONS = 'options'
export const ALL = '*'
