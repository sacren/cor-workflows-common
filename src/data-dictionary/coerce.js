import _ from 'lodash'
import CoercionNotSupportedError from './coercion-not-supported-error'

import * as booleanDataType from './data-types/boolean'
import * as userDataType from './data-types/user'

const DATA_TYPES = [booleanDataType, userDataType]

export const COERCIONS = _.reduce(
  DATA_TYPES,
  (accumulator, dataType) =>
    _.set(accumulator, dataType.RETURN_TYPE, dataType.COERCIONS),
  {}
)

export function coerce (baseType, requestedCoerceType, value) {
  const coercionFn = _.get(
    COERCIONS,
    [baseType, requestedCoerceType],
    coercionNotSupported(baseType, requestedCoerceType)
  )
  return coercionFn(value)
}

export const coercionNotSupported = _.curry(
  function throwCoercionNotSupportedError (
    baseType,
    requestedCoerceType,
    value
  ) {
    throw new CoercionNotSupportedError(baseType, requestedCoerceType, value)
  }
)
