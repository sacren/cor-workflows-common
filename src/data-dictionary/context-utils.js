/* Copyright Â© 2016 Kuali, Inc. - All Rights Reserved
 * You may use and modify this code under the terms of the Kuali, Inc.
 * Pre-Release License Agreement. You may not distribute it.
 *
 * You should have received a copy of the Kuali, Inc. Pre-Release License
 * Agreement with this file. If not, please write to license@kuali.co.
 */
import { isArray, keyBy, last } from 'lodash'
import axios from 'axios'
import Context from './context'
import Root from './root'
import GlobalCategories from './global-categories'
import GlobalForms from './global-forms'
import GlobalGroups from './global-groups'
import GlobalUsers from './global-users'
import Category from './global-categories/category'
import Form from './global-forms/form'
import Group from './global-groups/group'
import User from './global-users/user'
import Role from './global-roles/role'
import FieldChecklist from './global-forms/field-checklist'
import FieldCoreGroupMultiselect
  from './global-forms/field-core-group-multiselect'
import FieldCoreGroupTypeahead from './global-forms/field-core-group-typeahead'
import FieldCoreSurveyGrid from './global-forms/field-core-survey-grid'
import FieldCoreUserMultiselect
  from './global-forms/field-core-user-multiselect'
import FieldCoreUserTypeahead from './global-forms/field-core-user-typeahead'
import FieldCountryDropDown from './global-forms/field-country-dropdown'
import FieldDatePicker from './global-forms/field-date-picker'
import FieldDropDown from './global-forms/field-drop-down'
import FieldFormsMultiselect from './global-forms/field-forms-multiselect'
import FieldFormsTypeahead from './global-forms/field-forms-typeahead'
import FieldLanguageDropDown from './global-forms/field-language-dropdown'
import FieldRadioButton from './global-forms/field-radio-button'
import FieldRichTextArea from './global-forms/field-rich-text-area'
import FieldStateProvinceDropDown
  from './global-forms/field-state-province-dropdown'
import FieldTextArea from './global-forms/field-text-area'
import FieldTextInput from './global-forms/field-text-input'
import TextInput from './global-inputs/text-input'
import NumericInput from './global-inputs/numeric-input'
import categoryAPI from '../api/category'
import flowInstanceAPI from '../api/flow-insts'
import flowAPI from '../api/flow'
import formAPI from '../api/form'
import groupAPI from '../api/group'
import userAPI from '../api/user'
import { ALL } from './return-types'
import { contexts as additionalContexts, apis as additionalApis } from '../extensions'

const i18n = {
  UNSUPPORTED_TYPE: 'Cannot getContextType of a non-deflated context'
}

export const contexts = [
  Root,
  GlobalCategories,
  GlobalForms,
  GlobalGroups,
  GlobalUsers,
  Category,
  Form,
  Group,
  User,
  Role,
  FieldChecklist,
  FieldCoreGroupMultiselect,
  FieldCoreGroupTypeahead,
  FieldCoreSurveyGrid,
  FieldCoreUserMultiselect,
  FieldCoreUserTypeahead,
  FieldCountryDropDown,
  FieldDatePicker,
  FieldDropDown,
  FieldFormsMultiselect,
  FieldFormsTypeahead,
  FieldLanguageDropDown,
  FieldRadioButton,
  FieldRichTextArea,
  FieldStateProvinceDropDown,
  FieldTextArea,
  FieldTextInput,
  TextInput,
  NumericInput
].concat(additionalContexts)

export const apis = [
  categoryAPI,
  flowInstanceAPI,
  flowAPI,
  formAPI,
  groupAPI,
  userAPI
].concat(additionalApis)

export const DEFAULT_SCOPE = {
  axios,
  returnTypes: ALL
}

export class ContextUtil {
  static contextToArray = (context, array = []) => {
    if (!context) return array
    array.unshift(context)
    if (context.parent && context.parent.type !== 'root') {
      ContextUtil.contextToArray(context.parent, array)
    }
    return array
  }

  static getDeflatedContext (deflated) {
    // TODO: make the context handle providing basic info from a deflated array
    if (!isArray(deflated)) throw new Error(i18n.UNSUPPORTED_TYPE)
    return last(deflated)
  }

  static getDeflatedContextType (deflated) {
    if (!isArray(deflated)) throw new Error(i18n.UNSUPPORTED_TYPE)
    return ContextUtil.getDeflatedContext(deflated).type
  }

  static getDeflatedContextName (deflated) {
    if (!isArray(deflated)) throw new Error(i18n.UNSUPPORTED_TYPE)
    return ContextUtil.getDeflatedContext(deflated).name
  }

  constructor (scope) {
    this.contexts = contexts
    this.contextMap = keyBy(contexts, 'type')
    this.Root = Root
    this.scope = Object.assign({}, DEFAULT_SCOPE, scope)
    this.apiSet = apis
    this.apis = this.initAPIs()
    this.initRoot()
  }

  initAPIs () {
    const constructed = this.apiSet.map(Api => new Api(this.scope.axios))
    return keyBy(constructed, c => c.constructor.API_KEY)
  }

  initRoot (returnTypes) {
    this.root = new this.Root(
      null,
      returnTypes || this.scope.returnTypes,
      null,
      this
    )
    return this.root
  }

  getRoot () {
    return this.root || this.initRoot()
  }

  addContext (Context) {
    this.contexts.push(Context)
    this.contextMap = keyBy(this.contexts, 'type')
  }

  addAPI (Api) {
    this.apiSet.push(Api)
    this.apis = this.initAPIs()
  }

  deflate (value) {
    if (!(value instanceof Context)) return value
    return value.deflate()
  }

  /**
   * Recontitute contexts that have been stored
   * @param {Array} deflated The array of deflated values
   */
  async inflate (deflated, returnTypes = '*') {
    const { scope } = this
    const promises = deflated.map(d => {
      if (d.requiresParent) return
      return this.contextMap[d.type].inflate(this, d, undefined, scope)
    })
    const datas = await Promise.all(promises)
    const contexts = []
    for (let i = 0; i < datas.length; i++) {
      const defd = deflated[i]
      let data = datas[i]
      const Context = this.contextMap[defd.type]
      const parent = i > 0 ? contexts[i - 1] : undefined
      if (defd.requiresParent) {
        data = datas[i] = await Context.inflate(this, defd, parent, scope)
      }
      const ctx = Context.type === this.Root.type
        ? this.root || new Root(parent, returnTypes, data, this)
        : new Context(parent, returnTypes, data, this)
      contexts[i] = ctx
    }
    return contexts[contexts.length - 1]
  }

  getDeflatedContextType (deflated) {
    if (!isArray(deflated)) throw new Error(i18n.UNSUPPORTED_TYPE)
    return last(deflated).type
  }

  getDeflatedContextName (deflated) {
    if (!isArray(deflated)) throw new Error(i18n.UNSUPPORTED_TYPE)
    return last(deflated).name
  }

  getRootParent (value) {
    if (!value) return undefined
    const parent = value.getParent()
    if (parent) return this.getRootParent(parent)
    return value.type === this.Root.type ? value : undefined
  }
}

const contextUtil = new ContextUtil()
export default contextUtil
