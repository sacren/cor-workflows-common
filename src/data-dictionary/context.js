/* Copyright Â© 2016 Kuali, Inc. - All Rights Reserved
 * You may use and modify this code under the terms of the Kuali, Inc.
 * Pre-Release License Agreement. You may not distribute it.
 *
 * You should have received a copy of the Kuali, Inc. Pre-Release License
 * Agreement with this file. If not, please write to license@kuali.co.
 */
import { isArray, isEqual } from 'lodash'
import { ALL } from './return-types'
import hash from 'shorthash'
import {
  CONTAINS,
  DOES_NOT_CONTAIN,
  IS,
  IS_NOT,
  BEGINS_WITH,
  ENDS_WITH,
  DOES_NOT_BEGIN_WITH,
  DOES_NOT_END_WITH,
  IS_EMPTY,
  IS_NOT_EMPTY
} from './operators'

const DEFAULTS = {
  returnTypes: '*',
  filter: { limit: 10 }
}

/**
 * Base Context Class
 */
export default class Context {
  static global = true
  static type = 'base-context'
  static treatAsType = undefined
  static displayName = 'Data Source'
  static returnTypes = ALL
  static matchTypes = false
  static preferredOperators = [
    CONTAINS,
    DOES_NOT_CONTAIN,
    IS,
    IS_NOT,
    BEGINS_WITH,
    ENDS_WITH,
    DOES_NOT_BEGIN_WITH,
    DOES_NOT_END_WITH,
    IS_EMPTY,
    IS_NOT_EMPTY
  ]

  /**
   * Not yet implemented
   * @abstract
   */
  static async inflate (deflated) {
    throw new Error('Not yet implemented')
  }

  /**
   * Construct new Context
   * @param {Array|string} returnTypes The array of return types or '*' for all
   * @param {Object} parent The parent context
   * @param {Object} data Any data that should be available
   */
  constructor (parent, returnTypes = DEFAULTS['returnTypes'], data, ctx) {
    const {
      type,
      typeLabel,
      treatAsType,
      displayName,
      matchTypes,
      preferredOperators
    } = this.constructor
    this.type = type
    this.treatAsType = treatAsType
    this.name = displayName
    this.secondaryName = typeLabel
    this.match = matchTypes
    this.preferredOperators = preferredOperators || []
    this.parent = parent
    this.returnTypes = returnTypes
    this.data = data
    this.ctx = ctx
  }

  getParent () {
    return this.parent
  }

  /**
   * Not yet implemented
   * @abstract
   * @param {dunno} filter
   */
  async getChildren (filter) {
    throw new Error('Not yet implemented')
  }

  async getData () {
    return this.data
  }

  setData (data) {
    this.data = data
  }

  isLeaf () {
    return false
  }

  isEqual (ctx) {
    return isEqual(this, ctx)
  }

  matches (match) {
    if (!this.match || !match) return false
    if (match === ALL) return true
    const thisMatch = isArray(this.match) ? this.match : [this.match]
    const thatMatch = isArray(match) ? match : [match]
    return thisMatch.some(tm => thatMatch.includes(tm)) || false
  }

  /**
   * Not yet implemented
   * @abstract
   */
  deflate () {
    throw new Error('Not yet implemented')
  }

  /**
   * Not yet implemented
   * @abstract
   */
  async getValue (valueMap = {}) {
    const { parent } = this
    if (parent) await parent.getValue(valueMap)
  }

  toString () {
    return this.parent ? `${this.parent.toString()} > ${this.name}` : this.name
  }

  toShortString () {
    return this.name
  }

  hash () {
    return this.parent
      ? `${this.parent.hash()}:${this.shortHash()}`
      : this.shortHash()
  }

  shortHash () {
    return `${hash.unique(this.type)}-${hash.unique(this.name)}`
  }

  // --- Utility Functions ---

  /**
   * Return the filtered set of passed contexts
   */
  reduceByReturnType (contexts) {
    return contexts.filter(Context => {
      return (
        Context.returnTypes === '*' ||
        Context.returnTypes.some(type => this.returnTypes.includes(type))
      )
    })
  }
}
