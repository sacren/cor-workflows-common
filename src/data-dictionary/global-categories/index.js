/* Copyright Â© 2016 Kuali, Inc. - All Rights Reserved
 * You may use and modify this code under the terms of the Kuali, Inc.
 * Pre-Release License Agreement. You may not distribute it.
 *
 * You should have received a copy of the Kuali, Inc. Pre-Release License
 * Agreement with this file. If not, please write to license@kuali.co.
 */
import Context from '../context'
import Category from './category'
import { CATEGORY, GROUP, ROLE, TEXT, USER } from '../return-types'

export default class GlobalCategories extends Context {
  static global = true
  static type = 'global-categories'
  static displayName = 'Categories'
  static treatAsType = undefined
  static returnTypes = [CATEGORY, GROUP, ROLE, TEXT, USER]

  static async inflate () {}

  async getChildren (filter) {
    if (!this.ctx) throw new Error('no ctx')
    const categories = await this.ctx.apis.categories.list(filter)
    return categories.map(
      category => new Category(this, this.returnTypes, category, this.ctx)
    )
  }

  deflate (valueList = []) {
    const { parent, type, name } = this
    if (parent) parent.deflate(valueList)
    valueList.push({ type, name, requiresParent: false })
    return valueList
  }
}
