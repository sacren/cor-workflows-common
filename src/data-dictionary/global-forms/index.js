/* Copyright Â© 2016 Kuali, Inc. - All Rights Reserved
 * You may use and modify this code under the terms of the Kuali, Inc.
 * Pre-Release License Agreement. You may not distribute it.
 *
 * You should have received a copy of the Kuali, Inc. Pre-Release License
 * Agreement with this file. If not, please write to license@kuali.co.
 */
import Context from '../context'
import Form from './form'
import { FORM } from '../return-types'

export default class GlobalForms extends Context {
  static global = true
  static type = 'global-forms'
  static displayName = 'Forms'
  static returnTypes = [FORM]

  static async inflate () {}

  async getChildren (filter) {
    if (!this.ctx) throw new Error('no ctx')
    let forms
    if (filter && filter.id) {
      forms = [await this.ctx.apis.forms.getForm(filter)]
    } else {
      forms = await this.ctx.apis.forms.list(filter)
    }
    const formCtxs = forms.map(
      form => new Form(this, this.returnTypes, form, this.ctx)
    )
    return formCtxs
  }

  deflate (valueList = []) {
    const { parent, type, name } = this
    if (parent) parent.deflate(valueList)
    valueList.push({ type, name, requiresParent: false })
    return valueList
  }
}
