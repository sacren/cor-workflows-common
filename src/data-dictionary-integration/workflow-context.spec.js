/* Copyright Â© 2016 Kuali, Inc. - All Rights Reserved
 * You may use and modify this code under the terms of the Kuali, Inc.
 * Pre-Release License Agreement. You may not distribute it.
 *
 * You should have received a copy of the Kuali, Inc. Pre-Release License
 * Agreement with this file. If not, please write to license@kuali.co.
 */

import WorkflowContext from './workflow-context'
import Ancestry from '../ancestry-util'
import { ALL } from '../data-dictionary/return-types'
import Context from '../data-dictionary/context'

const workflowData = function (ctxData) {
  return {
    ancestry: {},
    stepId: 1,
    context: new Context(null, null, ctxData)
  }
}

describe('Workflow Context', () => {
  it('should inflate both its ancestry and context', async () => {
    Ancestry.inflate = jest.fn().mockReturnValue({ name: 'my ancestry' })
    const context = {
      scope: {
        axios: '123'
      },
      inflate: jest.fn().mockReturnValue({ name: 'my context' })
    }
    const deflated = {
      ancestry: [1, 2, 3],
      context: {},
      stepId: '123'
    }
    const inflation = await WorkflowContext.inflate(
      context,
      deflated,
      null,
      null
    )
    expect(inflation).toMatchObject({
      ancestry: { name: 'my ancestry' },
      context: { name: 'my context' },
      stepId: '123'
    })
  })

  describe('Construction', () => {
    it('should throw an error if no data is provided', () => {
      expect(() => {
        // eslint-disable-next-line
        new WorkflowContext()
      }).toThrowErrorMatchingSnapshot()
    })

    it('should throw and error if incorrect is provided', () => {
      // no ancestry
      expect(() => {
        // eslint-disable-next-line
        new WorkflowContext(null, null, {})
      }).toThrowErrorMatchingSnapshot()

      // no stepId
      expect(() => {
        // eslint-disable-next-line
        new WorkflowContext(null, null, { ancestry: {} })
      }).toThrowErrorMatchingSnapshot()

      // no context
      expect(() => {
        // eslint-disable-next-line
        new WorkflowContext(null, null, { ancestry: {}, stepId: 1 })
      }).toThrowErrorMatchingSnapshot()

      // wrong type of context
      expect(() => {
        // eslint-disable-next-line
        new WorkflowContext(null, null, {
          ancestry: {},
          stepId: 1,
          context: {}
        })
      }).toThrowErrorMatchingSnapshot()
    })
  })

  it('should not throw an and error and set properties with the correct data', () => {
    const contextData = { name: 'context data' }
    const data = workflowData(contextData)
    const wf = new WorkflowContext(null, null, data)

    expect(wf.ancestry).toMatchObject({})
    expect(wf.stepId).toEqual(1)
    expect(wf.context).toEqual(data.context)
    expect(wf.name).toEqual(Context.displayName)
    expect(wf.data).toEqual(contextData)
  })

  it('should get children from its context and update their parent and return types', async () => {
    const wfData = workflowData()
    const contextData = [{ name: 'child1' }, { name: 'child2' }]
    wfData.context.getChildren = jest.fn().mockReturnValue(contextData)
    const wf = new WorkflowContext(null, ALL, wfData)
    const children = await wf.getChildren()
    expect(children).toHaveLength(contextData.length)
    children.forEach((child, i) => {
      expect(child.parent).toEqual(wf)
      expect(child.returnTypes).toEqual(wf.returnTypes)
    })
  })

  describe('getData', () => {
    let wfData, wf
    beforeEach(() => {
      wfData = workflowData()
      wf = new WorkflowContext(null, null, wfData)
    })

    it('should call the parent getValue if parent is available and get the formfill value', async () => {
      let passedMap
      const parent = {
        getValue: jest.fn().mockImplementation(map => {
          passedMap = { ...map }
        })
      }
      wf.getFormfillValue = jest.fn()
      wf.parent = parent
      wf.context.type = 'formfill'
      await wf.getValue({})
      expect(passedMap).toMatchObject({})
      expect(wf.getFormfillValue).toHaveBeenCalledWith(wfData.context, {})
    })

    it('should throw an error if its context tyep is not formfill', async () => {
      await expect(wf.getValue()).toThrowErrorMatchingSnapshot()
    })
  })

  describe('Deflate', () => {
    let wfData, ctx
    beforeEach(() => {
      wfData = workflowData()
      ctx = {
        deflate: jest.fn().mockReturnValue({ name: 'deflatedContext' })
      }
      wfData.ancestry.deflate = jest
        .fn()
        .mockReturnValue({ name: 'deflatedAncestry' })
    })

    it('calls its parent deflate if present', () => {
      let parentDeflateList
      const parent = {
        deflate: jest.fn().mockImplementation(list => {
          parentDeflateList = [...list]
        })
      }
      const wf = new WorkflowContext(parent, null, wfData, ctx)
      wf.deflate()
      expect(parentDeflateList).toMatchObject([])
    })

    it('should deflate the pareht, context, and ancestry and update the value list', () => {
      const wf = new WorkflowContext(null, null, wfData, ctx)
      const deflation = wf.deflate()
      expect(ctx.deflate).toHaveBeenCalledWith(wfData.context)
      expect(wfData.ancestry.deflate).toHaveBeenCalled()
      expect(deflation).toMatchObject([
        {
          context: { name: 'deflatedContext' },
          name: wfData.context.name,
          ancestry: { name: 'deflatedAncestry' },
          stepId: 1,
          type: WorkflowContext.type,
          requiresParent: false
        }
      ])
    })
  })

  describe('getFormfillValue', () => {
    it('should perform an elaborate stitching of data as a return value', async () => {
      const formResponse = { name: 'i am form' }
      const schemaResponse = { name: 'i am schema' }
      const documentResponse = { name: 'i am document' }
      const ctx = {
        apis: {
          forms: {
            getForm: jest.fn().mockReturnValue(formResponse),
            getSchema: jest.fn().mockReturnValue(schemaResponse),
            getDocument: jest.fn().mockReturnValue(documentResponse)
          }
        }
      }
      const wf = new WorkflowContext(null, null, workflowData(), ctx)
      const definitionStep = { _id: 999 }
      const instanceStep = { meta: { form: { _id: 888 } } }
      wf.findDefinitionStep = jest
        .fn()
        .mockReturnValue({ step: definitionStep })
      wf.findInstanceStep = jest.fn().mockReturnValue(instanceStep)

      const formContext = { data: { id: '123' } }
      const valueMap = {
        definition: {},
        instance: { name: 'instance' },
        instances: { name: 'instances' }
      }
      const value = await wf.getFormfillValue(formContext, valueMap)
      expect(wf.findDefinitionStep).toHaveBeenCalledWith(
        formContext.data.id,
        valueMap.definition
      )
      expect(wf.findInstanceStep).toHaveBeenCalledWith(
        definitionStep._id.toString(),
        valueMap.instance,
        valueMap.instances
      )
      expect(ctx.apis.forms.getForm).toHaveBeenCalledWith({
        id: formContext.data.id
      })
      expect(ctx.apis.forms.getSchema).toHaveBeenCalledWith({
        id: formContext.data.id
      })
      expect(ctx.apis.forms.getDocument).toHaveBeenCalledWith(
        instanceStep.meta.form._id
      )
      const data = {
        container: formResponse,
        schema: schemaResponse,
        document: documentResponse,
        definitionStep,
        instanceStep
      }
      expect(valueMap.formfill).toMatchObject(data)
      expect(value).toMatchObject(data)
    })
  })

  describe('findInstanceStep', () => {
    let wf, instance
    beforeEach(() => {
      wf = new WorkflowContext(null, null, workflowData())
      instance = {
        steps: [
          { stepDefinitionId: 100 },
          { stepDefinitionId: 101 },
          { stepDefinitionId: 102 },
          { stepDefinitionId: 103 }
        ]
      }
    })

    it('should return undefined if no instance or instance steps are provided', () => {
      expect(wf.findInstanceStep()).toBeUndefined()
      expect(wf.findInstanceStep(null, {})).toBeUndefined()
    })

    it('should return an instance step if one if found', () => {
      const step = wf.findInstanceStep('102', instance)
      expect(step).toMatchObject(instance.steps[2])
    })

    it('should call findInstaceStep again with the next instance if it does not find a step', () => {
      const instances = {
        foo: {
          steps: [...instance.steps]
        }
      }
      instance.parent = { instance: 'foo' }
      const newStep = { stepDefinitionId: 104 }
      instances.foo.steps.push(newStep)
      const step = wf.findInstanceStep('104', instance, instances)
      expect(step).toMatchObject(newStep)
    })
  })

  describe('findDefinitionStep', () => {
    let steps, wf
    let createStep = function (type, formId) {
      const flowId = Math.floor(Math.random() * (10 - 0))
      return {
        flow: flowId,
        step: { type, meta: { form: [{ _id: formId }] } }
      }
    }

    beforeEach(() => {
      steps = [
        createStep('formfill', 222),
        createStep('notification', 333),
        createStep('conditional', 444),
        createStep('formfill', 555),
        createStep('parallel', 666)
      ]

      const wfData = workflowData()
      wfData.ancestry.forEachStep = jest.fn().mockImplementation(func => {
        steps.forEach(step => {
          func(step)
        })
      })
      wf = new WorkflowContext(null, null, wfData)
    })

    it('should be able to find a formfill step with a matching form id', () => {
      const step = steps[0]
      expect(wf.findDefinitionStep(222)).toMatchObject({
        flow: step.flow,
        step: step.step
      })
    })

    it('should not match a step with the correct id that is not a formfill', () => {
      expect(wf.findDefinitionStep(333)).toBeUndefined()
    })
  })
})
