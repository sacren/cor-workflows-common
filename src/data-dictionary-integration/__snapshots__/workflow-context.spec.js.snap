// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Workflow Context Construction should throw an error if no data is provided 1`] = `"Workflow Context expects data to have ancestry, stepId & context"`;

exports[`Workflow Context Construction should throw and error if incorrect is provided 1`] = `"Workflow Context expects data to have ancestry, stepId & context"`;

exports[`Workflow Context Construction should throw and error if incorrect is provided 2`] = `"Workflow Context expects data to have ancestry, stepId & context"`;

exports[`Workflow Context Construction should throw and error if incorrect is provided 3`] = `"Workflow Context expects data to have ancestry, stepId & context"`;

exports[`Workflow Context Construction should throw and error if incorrect is provided 4`] = `"Workflow Context expects data to have ancestry, stepId & context"`;

exports[`Workflow Context getData should throw an error if its context tyep is not formfill 1`] = `"received is not a function"`;
